### ISSUE TITLE
Document metadata.namespace behavior for webhooks

### ISSUE THREAD

**Issue by riking** [kubernetes/kubernetes#88282]
**What would you like to be added**:
Webhook authors should be aware that the metadata.namespace of submitted objects is not trustworthy in Mutating Webhooks, and may be blank in Validation Webhooks. The namespace needs to always be read out of the AdmissionReview object.

**Why is this needed**:
If a webhook author does not test their code with a modified kubectl default namespace, a validating webhook may incorrectly think a resource is being created in the `default` namespace.
Namespaces that mismatch between the URL and resource are not blocked by the apiserver until after the object has been sent to Mutating webhooks. A mutating webhook that reads the wrong namespace field may be vulnerable to information disclosure in error messages, or in extreme cases acting as a Confused Deputy if the webhook has side effects (please do not make a mutating webhook with side effects).

The correct fix for these problems is to fix the webhook code via developer education. Improve the docs so this is obvious.

/sig api-machinery
/sig docs
/area ecosystem

**Comment by sftim** [comment]
I looked at the page https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook and saw that it [currently](https://github.com/kubernetes/website/blob/dff2f7b9601470dc8383d883b23188ed450d89c2/content/en/docs/reference/access-authn-authz/admission-controllers.md) mentions side effects of a webhook without any cautionary note about this being a bad idea.

**Comment by riking** [comment]
> I looked at the page https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook and saw that it [currently](https://github.com/kubernetes/website/blob/dff2f7b9601470dc8383d883b23188ed450d89c2/content/en/docs/reference/access-authn-authz/admission-controllers.md) mentions side effects of a webhook without any cautionary note about this being a bad idea.

Yep, I think I had that paragraph in mind while writing the issue text.

**Comment by riking** [comment]
This section https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#best-practices-and-warnings would be an excellent place to put this warning.

**Comment by riking** [comment]
This section https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#best-practices-and-warnings would be an excellent place to put this warning.

To be clear, this is the issue:

> Namespaces that mismatch between the URL and resource are not blocked until after the object has been sent to Mutating webhooks.

/remove-lifecycle rotten

**Comment by liggitt** [comment]
> This is intended behavior because:
> 
> Mutating webhooks are, in theory, allowed to modify the namespace field.

This isn't accurate… the namespace the object ends up with must match the namespace of the URL it was submitted to.

**Comment by sftim** [comment]
@riking would you be willing to revise the issue description based on https://github.com/kubernetes/kubernetes/issues/88282#issuecomment-689213634 ?
